demoapps/passwordlogin/java/org/openyolo/demoapp/passwordlogin/LoginViewModel.java

```diff
@@ -0,0 +1,513 @@
+            // the following cases (which are enumerated here to demonstrate the possible
+            // responses), but for simplicity in this demo app we treat them all as requiring
+            // manual authentication by the user
+            case HintRetrieveResult.CODE_NO_HINTS_AVAILABLE:
+                Log.i(TAG, "Provider indicated there are no hints available");
+                break;
+            case HintRetrieveResult.CODE_USER_CANCELED:
+                Log.i(TAG, "User canceled the hint selection flow");
+                break;
+            case HintRetrieveResult.CODE_USER_REQUESTS_MANUAL_AUTH:
+                Log.i(TAG, "User requests manual authentication");
+                break;
+            case HintRetrieveResult.CODE_BAD_REQUEST:
+                Log.i(TAG, "Proivder indicates that the request was malformed");
+                break;
+            case HintRetrieveResult.CODE_UNKNOWN:
+            default:
+                Log.i(TAG, "An unknown error occurred in the hint retrieval flow");
+                break;
+        }
+
+        doManualAuthentication(
+                R.string.enter_email_prompt,
+                null);
+    }
+
+    /**
+     * Prompts the user to manually authenticate, with a context-specific message.
+     */
+    private void doManualAuthentication(
+            @StringRes int authPromptId,
+            @Nullable String userEmail) {
+        this.authPrompt.set(getResourceString(authPromptId));
+        this.email.set(userEmail != null ? userEmail : "");
+        this.showPasswordField.set(userEmail != null);
+        this.password.set("");
+        setLoaded();
+    }
+
+    /**
+     * Pushes handling of a sign in button click on the form to a background thread for processing.
+     */
+    @MainThread
+    public void signInButtonClicked(View view) {
+        getExecutor().execute(this::tryManualAuthentication);
+    }
+
+    /**
+     * Processes an attempt to manually authenticate.
+     */
+    private void tryManualAuthentication() {
+        String userEmail = this.email.get().trim();
+        String userPassword = this.password.get();
+
+        if (!Patterns.EMAIL_ADDRESS.matcher(userEmail).matches()) {
+            emailError.set(getResourceString(R.string.error_invalid_email));
+            return;
+        } else {
+            emailError.set("");
+        }
+
+        if (userPassword.length() < MIN_PASSWORD_LENGTH) {
+            passwordError.set(getResourceString(R.string.error_password_too_short));
+            return;
+        } else {
+            emailError.set("");
+        }
+
+        if (mUserDataSource.isExistingAccount(userEmail)) {
+            if (userPassword.isEmpty()) {
+                authPrompt.set(getResourceString(R.string.existing_account_enter_password));
+                showPasswordField.set(true);
+            } else if (mUserDataSource.authWithPassword(userEmail, userPassword)) {
```

Comment:
Nit: Avoid side effects in conditional expression. Similarly for if conditional below.

Reply:
Done.